### PR TITLE
Add TPC-H query 17 to TpchQueryBuilder

### PR DESCRIPTION
**Pull request by pramodsatya**
```
| # Num Threads/ Drivers | Velox(seconds) | DuckDB(seconds) |
|:----------------------:|:--------------:|:---------------:|
|            1           |      11.92     |      15.04      |
|            4           |      4.31      |       3.58      | 
|            8           |      3.27      |       3.31      | 
|           16           |      3.15      |       3.94      |
```

### CODE REVIEW

**Review by aditi-pandit — APPROVED** [review]
Thanks for this change !

**Comment by aditi-pandit** [comment]
There is a failure in the circleCI tests that doesn't look related to this change. Please follow up on it.

**Comment by mbasmanova** [comment]
@pramodsatya Velox seems to be a lot slower than DuckDB. Have you had a chance to investigate why?

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 1]
```diff
@@ -1383,6 +1385,73 @@ TpchPlan TpchQueryBuilder::getQ16Plan() const {
...
+              "",
+              {"l_quantity", "p_partkey", "l_extendedprice"})
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kLineitem, lineitemSelectedRowType, lineitemFileColumns)
+          .capturePlanNodeId(lineitemScanNodeIdSubQuery)
+          .partialAggregation({"l_partkey"}, {"avg(l_quantity) as avg_"})
+          .localPartition({})
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:1430]: This makes the processing go single-threaded. You should instead partition on l_partkey

```
.localPartition({"l_partkey"})
```

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 2]
```diff
@@ -1383,6 +1385,73 @@ TpchPlan TpchQueryBuilder::getQ16Plan() const {
...
+          .finalAggregation()
+          .hashJoin(
+              {"l_partkey"},
+              {"p_partkey"},
+              lineitemJoinPart,
+              "",
+              {"l_extendedprice", "l_quantity", "avg_"},
+              core::JoinType::kRight)
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:1438]: Why right join? Rows will null on the left side won't pass the filter that follows? Can this be an inner join with a filter evaluated as part of the join itself?

File: velox/exec/tests/utils/TpchQueryBuilder.cpp (hunk 1 above)
Comment (pramodsatya) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp]: Thanks for the suggestion, updated

File: velox/exec/tests/utils/TpchQueryBuilder.cpp (hunk 2 above)
Comment (pramodsatya) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp]: Removed the right join with the new plan as suggested, thanks!

**Comment by pramodsatya** [comment]
Thanks for the suggestion @mbasmanova, modified the query plan accordingly and the performance is now better than DuckDb. For the earlier plan, the average aggregate calculation was better in DuckDb although tablescan was faster in Velox; the optimized aggregation calculation has helped now.

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 3]
```diff
@@ -1383,6 +1385,85 @@ TpchPlan TpchQueryBuilder::getQ16Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ17Plan() const {
+  std::vector<std::string> lineitemColumns = {
+      "l_partkey", "l_extendedprice", "l_quantity"};
+  std::vector<std::string> partColumns = {
+      "p_partkey", "p_brand", "p_container"};
+
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:1394]: nit: drop 'Selected' for brevity: lineitemRowType and partRowType

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 4]
```diff
@@ -1383,6 +1385,85 @@ TpchPlan TpchQueryBuilder::getQ16Plan() const {
...
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  auto partSelectedRowType = getRowType(kPart, partColumns);
+  const auto& partFileColumns = getFileColumnNames(kPart);
+
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeId;
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:1400]: nit: drop 'Node' for brevity

nit: The code is a bit difficult to read. I understand there are 2 identical joins between lineitem and filtered part. Consider adding a lambda to produce that plan and fill in pair of scan node IDs. Also, make sure that lineitemScanId and partScanId point to nodes in one subquery and lineitemAggScanId and partAggScanId point to nodes in another subquery. As written, it seems that lineitemScanNodeId and partScanNodeIdSubQuery go together, which is confusing.

lineitemScanId, partScanId, lineitemAggScanId, partAggScanId

**Review by mbasmanova — APPROVED** [review]
@pramodsatya Looks good, but I wonder if there is a way to make the code a bit more readable.

File: velox/exec/tests/utils/TpchQueryBuilder.cpp (hunk 3 above)
Comment (pramodsatya) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp:1394]: Done, thanks!

File: velox/exec/tests/utils/TpchQueryBuilder.cpp (hunk 4 above)
Comment (pramodsatya) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp:1400]: Thanks for the suggestion, updated scanIds accordingly. Writing a lambda would require the optimal join ordering to be changed and the average aggregation would need to be calculated after the join for the subquery as well, which seems cumbersome.

**Comment by mbasmanova** [comment]
@pramodsatya Pramod, thank you for editing the code. The new version is much more readable.

**Comment by mbasmanova** [comment]
@pramodsatya Pramod, would you rebase onto latest main, so I can merge this?

**Comment by pramodsatya** [comment]
Rebased onto latest main, thanks!

**Pull request closed without merging**